components/timezone/src/ixdtf.rs

```diff
@@ -0,0 +1,424 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::{CustomTimeZone, CustomZonedDateTime, GmtOffset, InvalidOffsetError};
+use icu_calendar::{AnyCalendar, Iso, RangeError};
+use ixdtf::{parsers::records::UTCOffsetRecord, ParseError as IxdtfParseError};
+
+#[cfg(feature = "compiled_data")]
+use crate::TimeZoneIdMapper;
+#[cfg(feature = "compiled_data")]
+use icu_calendar::{Date, Time};
+#[cfg(feature = "compiled_data")]
+use ixdtf::parsers::{
+    records::{IxdtfParseRecord, TimeZoneRecord},
+    IxdtfParser,
+};
+
+/// The error type for parsing IXDTF syntax strings in `icu_timezone`.
+#[derive(Debug, PartialEq)]
+#[non_exhaustive]
+pub enum ParseError {
+    /// Syntax error for IXDTF string.
+    Syntax(IxdtfParseError),
+    /// Parsed record is out of valid date range.
+    Range(RangeError),
+    /// There was no time zone offset or IANA identifer found.
+    MissingTimeZone,
+    /// There were missing fields required to parse component.
+    MissingFields,
+    /// There were two offsets provided that were not consistent with each other.
+    InconsistentTimeZoneOffsets,
+    /// There was an invalid Offset.
+    InvalidOffsetError,
+    /// There was an invalid IANA identifier provided.
+    InvalidIanaIdentifier,
+    /// An unknown calendar was provided.
+    UnknownCalendar,
+}
+
+impl From<IxdtfParseError> for ParseError {
+    fn from(value: IxdtfParseError) -> Self {
+        Self::Syntax(value)
+    }
+}
+
+impl From<RangeError> for ParseError {
+    fn from(value: RangeError) -> Self {
+        Self::Range(value)
+    }
+}
+
+impl From<InvalidOffsetError> for ParseError {
+    fn from(_: InvalidOffsetError) -> Self {
+        Self::InvalidOffsetError
+    }
+}
+
+impl TryFrom<UTCOffsetRecord> for GmtOffset {
+    type Error = InvalidOffsetError;
+
+    fn try_from(value: UTCOffsetRecord) -> Result<Self, Self::Error> {
+        let hour_seconds = i32::from(value.hour) * 3600;
+        let minute_seconds = i32::from(value.minute) * 60;
+        Self::try_from_offset_seconds(
+            i32::from(value.sign as i8) * (hour_seconds + minute_seconds + i32::from(value.second)),
+        )
+    }
+}
```

Comment:
Nit, here and elsewhere: In https://github.com/unicode-org/icu4x/pull/5260#discussion_r1682820501 we decided not to publicly export these direct conversion functions, at least not until ixdtf 1.0. I think just keep them as private functions. But, if you think these are useful on their own, feel free to make a case for keeping them.

Reply:
I have no specific case for keeping them. As a side note, there are probably some places in this PR that could be streamlined a bit more having the `try_from_ixdtf_record` publicly available in `icu_calendar`. At the same time though, I'm not entirely convinced it's worth it to expose those methods.